edit-post/api/plugin.js

```diff
@@ -0,0 +1,87 @@
+/* eslint no-console: [ 'error', { allow: [ 'error' ] } ] */
+
+/**
+ * WordPress dependencies
+ */
+import { applyFilters } from '@wordpress/hooks';
+import { Component } from '@wordpress/element';
+
+/* External dependencies */
+import { isFunction, map } from 'lodash';
+
+/**
+ * Internal dependencies
+ */
+import { PluginContextProvider, withPluginContext } from './components/context';
+
+const plugins = {};
+
+/**
+ * Registers a plugin to the editor.
+ *
+ * @param {Object}   settings        The settings for this plugin.
+ * @param {string}   settings.name   The name of the plugin.
+ * @param {Function} settings.render The function that renders the plugin.
+ *
+ * @return {Object} The final plugin settings object.
+ */
+function registerPlugin( settings ) {
+	if ( typeof settings !== 'object' ) {
+		console.error(
+			'No settings object provided!'
+		);
+		return null;
+	}
+	if ( typeof settings.name !== 'string' ) {
+		console.error(
+			'Plugin names must be strings.'
+		);
+		return null;
+	}
+	if ( ! /^[a-z][a-z0-9-]*$/.test( settings.name ) ) {
+		console.error(
+			'Plugin names must include only lowercase alphanumeric characters or dashes, and start with a letter. Example: "my-plugin".'
+		);
+		return null;
+	}
+	if ( plugins[ settings.name ] ) {
+		console.error(
+			`Plugin "${ settings.name }" is already registered.`
+		);
+	}
+	if ( ! isFunction( settings.render ) ) {
+		console.error(
+			'The "render" property must be specified and must be a valid function.'
+		);
+		return null;
+	}
+
+	settings = applyFilters( 'editPost.registerPlugin', settings, settings.name );
+
+	return plugins[ settings.name ] = settings;
+}
+
+/**
+ * A component that renders all plugin fills in a hidden div.
+ */
+class PluginArea extends Component {
+	render() {
+		return (
+			<div id="plugin-fills" style={ { display: 'none' } }>
+				{ map( plugins, plugin => {
```

Comment:
I don't think it's a good idea to have it like this in the long run. It should be provided as a prop so the plugin area could rerender when a new plugin is registered. At the moment it won't rerender until it is going to be explicitly unmounted by the top component. It shouldn't be an issue for the time being as most of the plugins will register plugins before `edit-post` is loaded by the browser. We will have to fix it somehow.

Reply:
The most obvious choice would be to register plugins to the store and have this component listen to it.

Reply:
Yes, using the store seems like the best fit. My only concern is that plugin object contains non-serializable parts - in particular, the component. I don't know how strict about that we want to be. In fact, I don't think we would ever want to persist the list of registered plugins. We might give it a try. We could move `PluginArea` to its own file with that in place.